plugins/state_history_plugin/state_history_plugin.cpp

```diff
@@ -100,71 +126,74 @@ struct state_history_plugin_impl : std::enable_shared_from_this<state_history_pl
       return {};
    }
 
-   struct session : std::enable_shared_from_this<session> {
-      std::shared_ptr<state_history_plugin_impl> plugin;
-      std::unique_ptr<ws::stream<tcp::socket>>   socket_stream;
+   struct session_base {
+      virtual void send_update(const block_state_ptr& block_state) = 0;
+      virtual void close()                                         = 0;
+      virtual ~session_base() = default;
+      std::optional<get_blocks_request_v0>       current_request;
+   };
+
+
+   template <typename SocketType>
+   struct session : session_base, std::enable_shared_from_this<session<SocketType>> {
+     std::shared_ptr<state_history_plugin_impl>  plugin;
+      ws::stream<SocketType>                     socket_stream;
       bool                                       sending  = false;
       bool                                       sent_abi = false;
       std::vector<std::vector<char>>             send_queue;
-      std::optional<get_blocks_request_v0>       current_request;
       bool                                       need_to_send_update = false;
 
-      session(std::shared_ptr<state_history_plugin_impl> plugin)
-          : plugin(std::move(plugin)) {}
-
-      void start(tcp::socket socket) {
-         fc_ilog(_log, "incoming connection");
-         socket_stream = std::make_unique<ws::stream<tcp::socket>>(std::move(socket));
-         socket_stream->binary(true);
-         socket_stream->next_layer().set_option(boost::asio::ip::tcp::no_delay(true));
-         socket_stream->next_layer().set_option(boost::asio::socket_base::send_buffer_size(1024 * 1024));
-         socket_stream->next_layer().set_option(boost::asio::socket_base::receive_buffer_size(1024 * 1024));
-         socket_stream->async_accept([self = shared_from_this()](boost::system::error_code ec) {
+      session(std::shared_ptr<state_history_plugin_impl> plugin, SocketType socket)
+          : plugin(std::move(plugin)), socket_stream(std::move(socket)) {}
+
+      void start() {
+         ilog("incoming connection");
+         socket_stream.auto_fragment(false);
+         socket_stream.binary(true);
+         if constexpr (std::is_same_v<SocketType, tcp::socket>) {
+            socket_stream.next_layer().set_option(boost::asio::ip::tcp::no_delay(true));
+         }
+         socket_stream.next_layer().set_option(boost::asio::socket_base::send_buffer_size(1024 * 1024));
+         socket_stream.next_layer().set_option(boost::asio::socket_base::receive_buffer_size(1024 * 1024));
+        
+         socket_stream.async_accept([self = this->shared_from_this()](boost::system::error_code ec) {
```

Comment:
This codebase does not use explicit `this` pointers anywhere they are optional.

Reply:
Removing  them breaks compilation, must be there is a workaround I cant see... Thank you!